dora/src/sym.rs

```diff
@@ -44,118 +55,111 @@ impl SymTable {
     }
 
     pub fn get_class(&self, name: Name) -> Option<ClassId> {
-        self.get(name).and_then(|n| n.to_class())
+        self.get_type(name).and_then(|n| n.to_class())
     }
 
     pub fn get_const(&self, name: Name) -> Option<ConstId> {
-        self.get(name).and_then(|n| n.to_const())
+        self.get_term(name).and_then(|n| n.to_const())
     }
 
     pub fn get_fct(&self, name: Name) -> Option<FctId> {
-        self.get(name).and_then(|n| n.to_fct())
+        self.get_term(name).and_then(|n| n.to_fct())
     }
 
     pub fn get_struct(&self, name: Name) -> Option<StructId> {
-        self.get(name).and_then(|n| n.to_struct())
+        self.get_type(name).and_then(|n| n.to_struct())
     }
 
     pub fn get_trait(&self, name: Name) -> Option<TraitId> {
-        self.get(name).and_then(|n| n.to_trait())
+        self.get_type(name).and_then(|n| n.to_trait())
     }
 
     pub fn get_module(&self, name: Name) -> Option<ModuleId> {
-        self.get(name).and_then(|n| n.to_module())
+        self.get_term(name).and_then(|n| n.to_module())
     }
 
     pub fn get_enum(&self, name: Name) -> Option<EnumId> {
-        self.get(name).and_then(|n| n.to_enum())
+        self.get_type(name).and_then(|n| n.to_enum())
     }
 
     pub fn get_global(&self, name: Name) -> Option<GlobalId> {
-        self.get(name).and_then(|n| n.to_global())
+        self.get_term(name).and_then(|n| n.to_global())
     }
 
     pub fn get_var(&self, name: Name) -> Option<VarId> {
-        self.get(name).and_then(|n| n.to_var())
+        self.get_term(name).and_then(|n| n.to_var())
     }
 
-    pub fn insert(&mut self, name: Name, sym: Sym) -> Option<Sym> {
-        self.levels.last_mut().unwrap().insert(name, sym)
+    pub fn insert_type(&mut self, name: Name, sym: TypeSym) -> Option<TypeSym> {
+        self.levels.last_mut().unwrap().insert_type(name, sym)
+    }
+
+    pub fn insert_term(&mut self, name: Name, sym: TermSym) -> Option<TermSym> {
+        self.levels.last_mut().unwrap().insert_term(name, sym)
     }
 }
 
 #[derive(Debug)]
 pub struct SymLevel {
-    map: HashMap<Name, Sym>,
+    types: HashMap<Name, TypeSym>,
```

Comment:
While I like this split between `types` and `terms`, I am not convinced this is the right thing to do. AFAIU for example functions and classes live in separate maps, however we are not going to allow having both a function and class with the same name. The split doesn't really match language semantics.

Why not simply introduce separate maps for both modules and annotations instead?

Reply:
> AFAIU for example functions and classes live in separate maps, however we are not going to allow having both a function and class with the same name. The split doesn't really match language semantics.

I actually think it does – it's just that the way I'm checking it currently is clumsy:

When defining a `class Foo` (or struct) we should consider the name to be in both namespaces  (in _types_ because a class is a type; and in _terms_ because the class implicitly defines a  constructor that could be seen as a "static method").

Thinking about it, maybe I should just do exactly this – enter the name in both namespaces when encountering a class definition – then we can get rid of the special casing altogether.

> Why not simply introduce separate maps for both modules and annotations instead?

I think it's not a good idea to introduce ad-hoc namespaces – because these things tend to keep accumulating (Java has like 7 now) and make many things more complicated later on.

I believe the division between terms and types is a good choice based on languages that have done this (I mean both splitting things like this and the number of namespaces (2)).

Reply:
> > AFAIU for example functions and classes live in separate maps, however we are not going to allow having both a function and class with the same name. The split doesn't really match language semantics.
> 
> I actually think it does – it's just that the way I'm checking it currently is clumsy:
> 
> When defining a `class Foo` (or struct) we should consider the name to be in both namespaces (in _types_ because a class is a type; and in _terms_ because the class implicitly defines a constructor that could be seen as a "static method").
> 
> Thinking about it, maybe I should just do exactly this – enter the name in both namespaces when encountering a class definition – then we can get rid of the special casing altogether.

This sounds good, maybe then we can also mostly keep the current API and can handle the difference internally in `SymLevel`. For example I would probably still not split `Sym`, but depending on the `Sym` kind insert into a different map (or multiple).

> 
> > Why not simply introduce separate maps for both modules and annotations instead?
> 
> I think it's not a good idea to introduce ad-hoc namespaces – because these things tend to keep accumulating (Java has like 7 now) and make many things more complicated later on.
> 
> I believe the division between terms and types is a good choice based on languages that have done this (I mean both splitting things like this and the number of namespaces (2)).

I guess it depends on what kinds of overlaps you want to allow. But I agree that being more restrictive here is better.

Reply:
> This sounds good ...

Thanks for making me think about it, I think this solution will be way easier to read in the code.

> ... maybe then we can also mostly keep the current API and can handle the difference internally in SymLevel

I tried this before settling on splitting the symbols – I tried both having things as term-type-tuples and having Syms that are multiple things at once.

The outcomes of these approaches were not great complexity-wise.

I'd try to keep the split, because I believe that in the future we will have more places where we benefit from more matches that don't need to have catch-all case like in https://github.com/dinfuehr/dora/pull/192/files#diff-86db169ca86f2fbc58c5deec22c19fe0L325 or https://github.com/dinfuehr/dora/pull/192/files#diff-86db169ca86f2fbc58c5deec22c19fe0L455.

> I guess it depends on what kinds of overlaps you want to allow. But I agree that being more restrictive here is better.

Agreed, I'd try to be as restrictive as possible without it being an annoyance to write code.
There is a lot of benefit to be had of being able to say "I can look at this and know immediately to what this name refers to".

Reply:
I pushed changes to get rid of the special-handling in all places except when entering the class name as a constructor itself.

I think doing it this way is an improvemen, what do you think @dinfuehr?

Reply:
Indeed this is now much cleaner 👍

Reply:
Great!

(I'm hoping that I can get rid of the remaining special-hanbdling as well in the mid-to-long-term by treating the constructor more as a first-class element.)